File: 프로그래머스 삼각달팽이 -월간 코드 챌린지 시즌1-/Solution.java
class Solution {
    public static void main(String[] args) {
        Solution solution = new Solution();
        solution.solution(6);
    }

    public int[] solution(int n) {
        int[] answer = {};
        int[][] arr = new int[n][n];
        int max = n * (n + 1) / 2; // 총 개수
        int top = 0; //상단
        int left = 0; //좌측
        int bottom = n - 1; //하단
        int right = n - 1; //우측
        int value = 1;
        while (max >= value) {
            for (int i = top; i <= bottom; i++) { // 하단이동
                if (max < value) break;
                arr[i][left] = value++;
            }
            if (max < value) break;
            top++;
            left++;
            for (int j = left; j <= right; j++) { // 우측이동
                if (max < value) break;
                arr[bottom][j] = value++;
            }
            if (max < value) break;
            bottom--;
            right--;
            int index = right;
            for (int i = bottom; i >= top; i--) { //상단 이동
                if (max < value) {
                    break;
                }
                arr[i][index--] = value++;
            }
            top++;
            right--;
        }
        answer = new int[max];
        int index = 0;
        for (int i=0; i<n; i++){
            for (int j=0; j<=i; j++){
                answer[index++] =  arr[i][j];
            }
        }
        return answer;
    }
}

File: 프로그래머스 삼각달팽이 -월간 코드 챌린지 시즌1-/Solution.kt
package a

class Solution {
    fun solution(n: Int): IntArray {
        var answer: IntArray
        val arr = Array(n) { IntArray(n) }
        val max = n * (n + 1) / 2 // 총 개수
        var top = 0 //상단
        var left = 0 //좌측
        var bottom = n - 1 //하단
        var right = n - 1 //우측
        var value = 1
        while (max >= value) {
            for (i in top..bottom) { // 하단이동
                if (max < value) break
                arr[i][left] = value++
            }
            if (max < value) break
            top++
            left++
            for (j in left..right) { // 우측이동
                if (max < value) break
                arr[bottom][j] = value++
            }
            if (max < value) break
            bottom--
            right--
            var index = right
            for (i in bottom downTo top) { //상단 이동
                if (max < value) {
                    break
                }
                arr[i][index--] = value++
            }
            top++
            right--
        }
        answer = IntArray(max)
        var index = 0
        for (i in 0 until n) {
            for (j in 0..i) {
                answer[index++] = arr[i][j]
            }
        }
        return answer
    }
}